packages/framer-motion/src/animation/js/index.ts

```diff
@@ -0,0 +1,261 @@
+import { AnimationPlaybackControls } from "../animate"
+import { keyframes as keyframesGeneratorFactory } from "../generators/keyframes"
+import { spring } from "../generators/spring/index"
+import { inertia } from "../generators/inertia"
+import { AnimationState, KeyframeGenerator } from "../generators/types"
+import { DriverControls } from "./types"
+import { AnimationOptions } from "../types"
+import { frameloopDriver } from "./driver-frameloop"
+import { interpolate } from "../../utils/interpolate"
+import { clamp } from "../../utils/clamp"
+
+type GeneratorFactory = (
+    options: AnimationOptions<any>
+) => KeyframeGenerator<any>
+
+const types: { [key: string]: GeneratorFactory } = {
+    decay: inertia,
+    inertia,
+    tween: keyframesGeneratorFactory,
+    keyframes: keyframesGeneratorFactory,
+    spring,
+}
+
+function calculateDuration(generator: KeyframeGenerator<unknown>) {
+    let duration = 0
+    const timeStep = 50
+    let state = generator.next(duration)
+    while (!state.done && duration < 20000) {
```

Comment:
What is this constant? Can't we go higher than 20s? Maybe we should extract it into a const, and can we write it like `20_000` in typescript?

Reply:
(But will abstract to constant)